From fdde5e4f4a04ea629be0ad9aff6d8fd64642b393 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Sun, 3 Aug 2014 11:03:51 +0200
Subject: [PATCH 1017/1020] ASoC: imx-hdmi-dma: Fix md optimize subframe
 assembly.

Do not include unused bits of pcm_data in parity calculation and
avoid make better use of odd_ones().

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 sound/soc/fsl/imx-hdmi-dma.c |   27 +++++++++++++--------------
 1 file changed, 13 insertions(+), 14 deletions(-)

diff --git a/sound/soc/fsl/imx-hdmi-dma.c b/sound/soc/fsl/imx-hdmi-dma.c
index b320691..42e78df 100644
--- a/sound/soc/fsl/imx-hdmi-dma.c
+++ b/sound/soc/fsl/imx-hdmi-dma.c
@@ -229,31 +229,30 @@ static u32 hdmi_dma_add_frame_info(struct hdmi_dma_priv *priv,
 	union hdmi_audio_dma_data_t subframe;
 
 	subframe.U = 0;
-	iec_header.B.channel = subframe_idx;
-
-	/* fill b (start-of-block) */
-	subframe.B.b = (priv->frame_idx == 0) ? 1 : 0;
 
 	/* fill c (channel status) */
-	if (priv->frame_idx < 42)
-		subframe.B.c = (iec_header.U >> priv->frame_idx) & 0x1;
-	else
-		subframe.B.c = 0;
-
+	if (priv->frame_idx < 42) {
+		iec_header.B.channel = 
+			(iec_header.B.linear_pcm == 0) ? subframe_idx : 0;
+		subframe.B.c = iec_header.U >> priv->frame_idx;
+	}
+	
 	/* fill v (validity) */
 	subframe.B.v = iec_header.B.linear_pcm;
 
-	subframe.B.p = odd_ones(pcm_data);
-	subframe.B.p ^= subframe.B.c;
-	subframe.B.p ^= subframe.B.u;
-	subframe.B.p ^= subframe.B.v;
-
 	/* fill data */
 	if (priv->sample_bits == 16)
 		subframe.B.data = pcm_data << 8;
 	else
 		subframe.B.data = pcm_data;
 
+	/* fill p (parity) Note: Do not include b ! */
+	subframe.B.p = odd_ones(subframe.U);
+
+	/* fill b (start-of-block) */
+	if (priv->frame_idx == 0)
+		subframe.B.b = 1;
+
 	return subframe.U;
 }
 
-- 
1.7.9.5

